### PR TITLE
intel_adsp: ipc: pm action in busy state

### PR DESCRIPTION
**Pull request by tmleman**
Currently SOF has disabled CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE option and use pm_suspend_devices() to suspend and resume IPC device during D3 power flow. The pm_suspend_devices() function skips suspending devices that are busy. In very rare cases, the IPC device is busy during the power state transition, which results in the device not being restored during reboot. This happens when FW sends a message to the HOST and waits for ACK, and the HOST simultaneously sends a SET_DX message to the DSP. This suspend/resume logic in IPC driver does not work well when the system enters the D3 state because it is not a suspend state, but rather a power-off. IPC does not require suspending, only reinitialization when exiting D3. We cannot avoid this one missing ACK and it cannot block the DSP from turning off.

When FW receives a SET_DX message it checks whether it can enter the D3 state and then returns an error (via IPC) or calls the pm_state_force function. Success response is sent directly from power_down assembly and not via ipc driver. This is because after receiving the response, the HOST will turn off the DSP.

In order for the transition to D3 to take place, only the primary core can be active, all pipes must be stopped (and therefore all modules in FW). The only active thread at this time is the Idle thread. Driver on the host will not send another ipc because is still waiting for response. FW can try to send only two notification:

- FW exception: from this place there is no return to continue the power transition,
- log buffer status: skipped, they remain in the queue without being sent.

I'm moving pm_device_busy_clear(dev) from IRQ handler to intel_adsp_ipc_send_message function so the pending ACK does not block power transition.

### CODE REVIEW

**Comment by andyross** [comment]
A little confused?   This patch removes the busy check, which is an API intended to prevent power transitions.  It used to refuse, now it allows it.  But the commit message says "IPC is one of the devices that are always-on and
suspend step can be skipped".  Those don't seem to be compatible statements?  Is IPC doing PM or not?

Maybe you want to do a full revert of this instead? https://github.com/zephyrproject-rtos/zephyr/commit/23b3cae1b1d91cd064d31cdeb78e3a6127b5051d

But assuming you do want power management of the IPC device, having it ignore the busy flag (which is a stanard framework API) seems like the wrong thing.  If it does it it should follow the rules and not allow state transitions when busy.

**Comment by tmleman** [comment]
@andyross please look again. I improved the commit description, explaining exactly what I was doing and why.

**Comment by andyross** [comment]
> This happens when FW sends a message to the HOST and waits for ACK, and the HOST simultaneously sends a SET_DX message to the DSP.

That sounds like a race condition in the protocol, can't you handle it there by e.g. synthesizing a reply and locking out new incoming commands?  I'm really sensitive to adding weird implicit states to an already complicated state machine.  I mean, at some point you probably will want to move to normal/classic/traditional lazy evaluated power management.  And in that case this patch will have broken things because the IPC device will happily suspend while trasactions are in flight!

**Comment by tmleman** [comment]
@andyross I think I managed to find a less controversial solution.

For now, it is only to fix/workaround the existing problem. Later we should rethink the power management for the IPC device. In my opinion, it's really unnecessary. After HOST sends SET_DX there will be no other messages. There is only one notification from the DSP side that the FW may want to send and if this happens, we just ignore it.

**Review by jxstelter — APPROVED** [review]

**Review by softwarecki — APPROVED** [review]

File: soc/xtensa/intel_adsp/common/ipc.c [hunk 1]
```diff
@@ -82,7 +82,6 @@ void z_intel_adsp_ipc_isr(const void *devarg)
 		}
```
Comment (kv2019i) [inline soc/xtensa/intel_adsp/common/ipc.c:82]: A side-note: While I find the detailed git commit description very useful as a SOF developer, I think it has a bit too much SOF/app specific knowledge (like SET_DX message) that will be hard to follow in context of Zephyr. Being too verbose is probably not too big of a problem (especially as most people working on this specific driver will use it with SOF), but as a general rule it would be good to describe the problem without using SOF/application terms, and maybe add the mapping to SOF in a separate paragraph in the commit message.

File: soc/xtensa/intel_adsp/common/ipc.c [hunk 2]
```diff
@@ -163,6 +162,7 @@ int intel_adsp_ipc_send_message(const struct device *dev,
 	config->regs->idd = ext_data;
 	config->regs->idr = data | INTEL_ADSP_IPC_BUSY;
 	k_spin_unlock(&devdata->lock, key);
+	pm_device_busy_clear(dev);
```
Comment (kv2019i) [inline soc/xtensa/intel_adsp/common/ipc.c:165]: Should we clear the busy bit also for -EBUSY on L157 ? The old implementation is a bit suspect as well, but I think the expected flow is that the ack (in irq) will clear the bit in case send returns -EBUSY. If we now change the busy flag clearing to happen here synchronously, also the EBUSY handling needs to be changed.

**Review by kv2019i — COMMENTED** [review]
A few comments inline. I do think the approach seems sane. Keeping the PM busy set for duration of a protocol exchange does not seem correct and should be something handled in upper layers. E.g. fact the a notification (of e.g. panic) will not necessarily get a response, is not something we should hardcode in the IPC driver, but rather in the upper level entity that implements the protocol logic.

Not sure if setting/clearing the PM bit has any functional purpose in the driver, but if we want to proceed piecemeal, I can sign off this approach. But if we keep setting/clearing, we need to be consistent and handle all cases (see inline comment).

**Comment by nashif** [comment]
@ceolin  please review

**Review by ceolin — APPROVED** [review]
